apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/Keys.cs

```diff
@@ -226,45 +219,44 @@ public sealed class KeyRange
             EndInclusive = endInclusive;
         }
 
-        /// <inheritdoc/>
-        public override bool Equals(object other)
-            => other is KeyRange range && Protobuf.Equals(range.Protobuf);
-
-        /// <inheritdoc/>
-        public override int GetHashCode() => Protobuf.GetHashCode();
-
-        /// <inheritdoc/>
-        public override string ToString()
+        /// <summary>
+        /// Returns the string representation of this KeyRange using the type conversion options specified in 
+        /// <see cref="SpannerConnectionStringBuilder"/>. The string representation is only for diagnostic purposes
+        /// and the implementation may change over time.
+        /// </summary>
+        /// <param name="builder">The <see cref="SpannerConnectionStringBuilder"/> instance.</param>
+        /// <returns>A string that represents the current KeyRange.</returns>
+        public string ToString(SpannerConnectionStringBuilder builder)
         {
-            var builder = new StringBuilder();
-            builder.Append(BeginInclusive ? '[' : '(');
-            builder.Append(BeginAt.KeyParts);
-            builder.Append(", ");
-            builder.Append(End.KeyParts);
-            builder.Append(EndInclusive ? ']' : ')');
-
-            return builder.ToString();
+            var stringBuilder = new StringBuilder();
```

Comment:
We could easily make this an expression-bodied member - again, that's later refactoring.